driver-core/src/main/java/com/datastax/driver/core/PagingState.java

```diff
@@ -0,0 +1,158 @@
+    }
+
+    PagingState(ByteBuffer pagingState, ByteBuffer hashDigest) {
+        this.pagingState = pagingState;
+        this.hashDigest = hashDigest;
+    }
+
+    private byte[] generatePagingStateHash(Statement statement) {
+        byte[] digest;
+        ByteBuffer[] values;
+        MessageDigest md;
+
+        assert !(statement instanceof BatchStatement);
+
+        try {
+            md = MessageDigest.getInstance("MD5");
+            if (statement instanceof BoundStatement) {
+                BoundStatement bs = ((BoundStatement)statement);
+                md.update(bs.preparedStatement().getQueryString().getBytes());
+                values = bs.values;
+            } else {
+                //it is a RegularStatement since Batch stataments are not allowed
+                RegularStatement rs = (RegularStatement)statement;
+                md.update(rs.getQueryString().getBytes());
+                values = rs.getValues();
+            }
+            if (values != null) {
+                for (ByteBuffer value : values) {
+                    md.update(value.duplicate());
+                }
+            }
+            md.update(this.pagingState.duplicate());
+            digest = md.digest();
+
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("MD5 doesn't seem to be available on this JVM", e);
+        }
+        return digest;
+    }
+
+    boolean matches(Statement statement) {
+        ByteBuffer hashToCheck = this.hashDigest.duplicate();
+        byte[] toTest;
+        byte[] original = new byte[hashToCheck.remaining()];
+        hashToCheck.get(original);
+        toTest = this.generatePagingStateHash(statement);
+
+        return Arrays.equals(original, toTest);
+    }
+
+    /**
+     * Return the hash digest generated after the creation of the
+     * PagingState object.
+     *
+     * @return the MD5 hash generated wrapped in a {@link java.nio.ByteBuffer}.
+     */
+    public ByteBuffer getHashDigest() {
+        return this.hashDigest;
+    }
+
+    /**
+     * Return the paging state buffer returned on the result of a query.
+     *
+     * @return the paging state buffer in a {@link java.nio.ByteBuffer} object.
+     */
+    public ByteBuffer getRawState() {
+        return this.pagingState;
+    }
+
+    /**
+     * Export the paging state and its MD5 hash in a string with the
+     * two values separated by a "_".
+     *
+     * @return the string of the two values separated by a "_".
+     */
+    @Override
+    public String toString() {
+        return Bytes.toRawHexString(this.pagingState) + "_" + Bytes.toRawHexString(this.hashDigest);
+    }
+
+    /**
+     * Create a PagingState object from a string previously saved of a PagingState
+     * object. As for the construction of a paging state object, the values must match
+     * the same statement (and potential parameters) since a hash verification is going
+     * to be made from the string.
```

Comment:
This is a bit too specific, the hash verification will be made later. I think the only thing that matters is that the value that gets passed in must come from a previous invocation of `toString` on a `PagingState` instance.